dasharo-security/measured-boot.robot

```diff
@@ -18,25 +18,376 @@ Resource            ../keys.robot
 # - document which setup/teardown keywords to use and what are they doing
 # - go threough them and make sure they are doing what the name suggest (not
 # exactly the case right now)
-Suite Setup         Run Keyword
-...                     Prepare Test Suite
+Suite Setup         Measured Boot Suite Setup
 Suite Teardown      Run Keyword
 ...                     Log Out And Close Connection
 
 
+*** Variables ***
+${PCRS_TO_CHECK}=       [0-9]|14
+
+
 *** Test Cases ***
 MBO001.001 Measured Boot support (Ubuntu 20.04)
     [Documentation]    Check whether Measured Boot is functional and
-    ...    measurements are stored into the TPM.a.
-    Skip If    not ${MEASURED_BOOT_SUPPORT}    MBO001.001 not supported
-    Skip If    not ${TESTS_IN_UBUNTU_SUPPORT}    MBO001.001 not supported
+    ...    measurements are stored into the TPM.
+    Execute Command In Terminal    shopt -s extglob
+    ${hashes}=    Execute Command In Terminal
+    ...    grep . /sys/class/tpm/tpm0/pcr-sha@(1|256)/[0-3]
+    Should Not Contain Any    ${hashes}
+    ...    0000000000000000000000000000000000000000
+    ...    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
+    ...    No such file or directory
+    ...    error
+    ...    ignore_case=${TRUE}
+
+MBO002.001 Check if event log PCRs match actual values (Ubuntu 22.04)
+    [Documentation]    Check whether PCRs values calculated from event log match
+    ...    actual PCRs values
+    ${tpm2_eventlog}=    Execute Command In Terminal
+    ...    tpm2_eventlog /sys/kernel/security/tpm0/binary_bios_measurements
+    Should Not Contain    ${tpm2_eventlog}    ERROR: Unable to run tpm2_eventlog
+    FOR    ${algo}    IN    sha1    sha256
+        ${eventlog_pcrs}=    Get PCRs From Eventlog    ${tpm2_eventlog}    ${algo}
+        FOR    ${pcr_element}    IN    @{eventlog_pcrs}
+            ${pcr}    ${hash}=    Split String    ${pcr_element}    separator=:
+            ${sha_hash}=    Execute Command In Terminal
+            ...    cat /sys/class/tpm/tpm0/pcr-${algo}/${pcr}
```

Comment:
Commands here and in other places are specific to TPM2.  Are we intentionally not checking TPM1?

Reply:
At least on QEMU `tpm2_eventlog` works with TPM 1.2 `binary_bios_measurements`. It would need to be tested on real platform with TPM 1.2 to be sure.

Reply:
I checked. In trenchboot:

```shell
root@tb:~# ls /sys/class/tpm/tpm0/pcr*
/sys/class/tpm/tpm0/pcrs

/sys/class/tpm/tpm0/pcr-sha1:
0   1   10  11  12  13  14  15  16  17  18  19  2   20  21  22  23  3   4   5   6   7   8   9
root@tb:~# cat /sys/class/tpm/tpm0/tpm_version_major
1
```

Reply:
In ubuntu it's the same:

```shell
ubuntu@3mdeb:~$ ls /sys/class/tpm/tpm0/pcr*
/sys/class/tpm/tpm0/pcrs

/sys/class/tpm/tpm0/pcr-sha1:
0  10  12  14  16  18  2   21  23  4  6  8
1  11  13  15  17  19  20  22  3   5  7  9
```

Reply:
My bad, for some reason I was sure there is only `.../pcrs` for TPM1.

Reply:
One more thing though: the loop tries to use `sha256`, won't it fail for TPM1?

Reply:
No, you can check in logs in this comment: https://github.com/Dasharo/open-source-firmware-validation/pull/326#issuecomment-2214214232
Platform had only sha256. Keyword will return empty list if no shaX values are in event log.